public/static/docs/api-reference/open.md

```diff
@@ -0,0 +1,139 @@
+  (such as a branch name, a tag, or a commit hash). `rev` only has an effect
+  when a URL is supplied as parameter to `repo`. If not supplied, it uses the
+  default Git revision, `HEAD`.
+
+- `remote` - (optional) name of the [DVC remote](/doc/command-reference/remote)
+  to fetch the target artifact from. If not supplied, the default depends on the
+  value of `repo`. The local cache is used when `repo` is the current working
+  directory (default value of `repo`). when `repo` is an external repository
+  URL, the default project remote is used.
+
+  > A `dvc.exceptions.NoRemoteError` is thrown if no `remote` is specified and
+  > the project has no default remote.
+
+- `mode` - (optional) mirrors the namesake parameter in builtin
+  [`open()`](https://docs.python.org/3/library/functions.html#open). Defaults to
+  `"r"` (read).
+
+- `encoding` - (optional) used to decode contents to a string. Mirrors the
+  namesake parameter in builtin `open()`. Defaults to `"utf-8"`.
+
+## Example: process XML file from an external DVC repository
+
+```py
+from xml.dom.minidom import parse
+
+import dvc.api
+
+with dvc.api.open(
+  "get-started/data.xml",
+  "https://github.com/iterative/dataset-registry"
+) as fd:
+  xmldom = parse(fd)
+  # ... Process elements
+```
+
+> See also `dvc.api.read` for a more direct way to read the complete contents of
+> a file <abbr>artifact</abbr>.
+
+## Example: use a file from the local cache
+
+In this case we don't supply a `repo` value, which means the current working
+directory will be tried instead, so make sure that the code is run from within a
+<abbr>DVC project</abbr>:
+
+```py
+import dvc.api
+
+with dvc.api.open('data/nlp/words.txt') as fd:
+  print(fd.name)
+```
+
+DVC will look for `data/nlp/words.txt` in the local cache of the
+<abbr>project</abbr>. (If it's not found there, the default
+[remote](/doc/command-reference/remote) will be tried.)
+
+The output of the script above should be something like
+`.dvc/cache/3a/01762e96060aa04a68345fbd910355` – the physical data location.
+
+## Example: process CSV file from a private repository
+
+For this we'll have to use the SSH URL to the Git repo (assuming the local
+[SSH credentials](https://help.github.com/en/github/authenticating-to-github/connecting-to-github-with-ssh)
+are configured locally):
+
+```py
+import csv
+import dvc.api
+
+with dvc.api.open(
+  "sea_ice.csv",
+  repo="git@github.com:iterative/df_sea_ice_no_header.git"
+) as fd:
+  reader = csv.reader(fd)
+  for row in reader:
+    # ... Process columns
+```
+
+## Example: stream file from a specific remote
+
+Sometimes we may want to chose the [remote](/doc/command-reference/remote) data
+source, for example to ensure that file streaming is enabled. This can be done
+by providing a `remote` argument:
+
+```py
+import dvc.api
+
+with dvc.api.open(
+  'model.pkl',
+  repo='https://github.com/example/dvc-repository'
+  remote='my-s3-bucket'
+) as fd:
+  for line in fd:
+    # ... Process lines
```

Comment:
`model.pkl` is a binary file, some reading it line by line is weird. If it's a `.pkl` then the only sensible thing is unserializing it. Other possible use cases:

```python
# load a json file
import json

with ...:
    data = json.load(fd)


# import csv as a pandas dataframe
import pandas as pd

with ...:
    df = pf.read_csv(fd)


# stream and process a text file line by line
import pandas as pd

with open("some.log", repo=...) as fd:
    for line in fd:
        match = re.search(r"user=(\w+)", line)
        # ...
```

Reply:
Yep, thanks Alex. Some of these examples were dummy for someone to review them and provide better ideas, so thanks! I'm using the `stream and process a text file line by line` instead, and adding a separate one for unserializing `model.pkl`. Please check.